### ISSUE TITLE
Super resolution using Deep CNN's

### ISSUE THREAD

**Issue by GilLevi** [vlfeat/matconvnet#69]
Hi,

I'm trying to reproduce the method described in the following paper: 
Dong, Chao, et al. "Learning a deep convolutional network for image super-resolution." Computer Vision–ECCV 2014. Springer International Publishing, 2014. 184-199.‏

The method (if I understood it correctly) is to use two conv layers where the second one contains filters of size 1x1xcx1, thus producing images of the same size as the input images.

For some reason, the loss does not seem decrees. I defined the loss functions as follows:

function Y = vl_nnRMSEloss(X,c,dzdy)

if nargin <= 2
a=(X-c).^2;
Y=sum(a(:));
Y=Y/size(X,4);

else
Y = +((X-c))*dzdy;
end

Below is my code for defining the net and training. I'll just mention first that I tried to add a NORM layer and change the learning rates and nothing seems to work. It seems that after each forward-backward step, the filters only become larger and larger, making the predictions larger and the loss actually increases and becomes steady around a certain value after a few iterations. 

imdb.images.data = imresize(imresize(data,0.5),2);
imdb.images.data_mean = mean(data,4);
imdb.images.labels = data ;
imdb.images.set=[ones(1,3*M/5), 2*ones(1,1*M/5) , 3*ones(1,1*M/5)];

imdb.images.data = single(imdb.images.data);
imdb.images.data_mean = single(imdb.images.data_mean);
imdb.images.labels = single(imdb.images.labels);

% Define network architecture
net.layers = {} ;
filters1=zeros(5,5,1,32);   
for ix=1:32
    filters1(:,:,:,ix)=fspecial('gaussian',5,randi(20));
end

sigma = 5;
size = 32;
x = linspace(-size / 2, size / 2, size);
gaussFilter = exp(-x .^ 2 / (2 * sigma ^ 2));
gaussFilter = gaussFilter / sum (gaussFilter);

filters2=zeros(1,1,32,1);
filters2(1,1,:,1) = gaussFilter;
filters1=single(filters1);
filters2=single(filters2);

% 1 conv1
net.layers{end+1} = struct('type', 'conv', ...
  'filters', filters1 , ...
  'biases', zeros(1, 32, 'single'), ...
  'filtersLearningRate', 0.00001, ...
  'biasesLearningRate', 0.00002, ...
  'stride', 1, ...
  'pad', 1) ;

% 2 conv2
net.layers{end+1} = struct('type', 'conv', ...
  'filters', filters2 , ...
  'biases', zeros(1, 1, 'single'), ...
  'filtersLearningRate', 1, ...
  'biasesLearningRate', 2, ...
  'stride', 1, ...
  'pad', 1) ;   

% 4 RELU
net.layers{end+1} = struct('type', 'relu') ;

% 4 RMSE loss
net.layers{end+1} = struct('type', 'RMSEloss') ;

opts.train.batchSize = 100 ;
opts.train.numEpochs = 20 ;
opts.train.continue = false ;
opts.train.useGpu = false ;
opts.train.learningRate = [0.001*ones(1, 12) 0.0001*ones(1,6) 0.00001] ;

[net, info] = cnn_train(net, imdb, @getBatch, ...
    opts.train, ...
    'val', find(imdb.images.set == 2)) ;

Any insight on what might seems to be the problem?

Thanks!
Gil

**Comment by bazilas** [comment]
You could try to normalize your output values to 0-1 and also subtract the mean from your input data. If it still explodes, try to reduce more the learning rate. Do you use the same architecture - parameters with the referred paper?

**Comment by GilLevi** [comment]
Thanks @bazilas , I got it working.

I use the same architecture, but not the same number of filters. 

Gil.

**Comment by dteney** [comment]
Hi Gil,
I had a quick look at the paper, I don't think your layers are correct. You first want a 9x9 conv, then a 1x1 conv to map 64-dim feature map to a 32-dim feature map, then "reconstruction" convolutions that are 5x5. And I don't think you want a RELU after the last ("reconstruction") convolutions. Here's how I would do it:

  net.layers{end+1} = struct('type', 'conv', 'filters', 0.01 * randn(9, 9, 3, 64, 'single'), 'biases', zeros(1, 64, 'single'), 'stride', 2, 'pad', 0, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
  net.layers{end+1} = struct('type', 'relu') ;

  net.layers{end+1} = struct('type', 'conv', 'filters', 0.01 * randn(1, 1, 64, 32, 'single'), 'biases', p.initialBias*ones(1, 32, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
  net.layers{end+1} = struct('type', 'relu') ;

  net.layers{end+1} = struct('type', 'conv', 'filters', 0 * randn(5, 5, 32, 1, 'single'), 'biases', p.initialBias*ones(1, 1, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;

  net.layers{end+1} = struct('type', 'l2loss') ;

**Comment by dteney** [comment]
Oh, and the padding for the middle layer should be 0, not 2.

**Comment by pengsun** [comment]
Just cut in you guys discussion for a minor question: is there any flaw in the MSE code by @GilLevi ?

> function Y = vl_nnRMSEloss(X,c,dzdy)

> if nargin <= 2
> a=(X-c).^2;
> Y=sum(a(:));
> Y=Y/size(X,4);

> else
> Y = +((X-c))*dzdy;
> end

I think the fprop() and bprop() pairing should be:
``` Matlab 
a = 0.5*(X-c).^2;
```
and
``` Matlab
Y = + ((X-c))*dzdy;
``` 
Otherwise it would confuse when setting the step size for SGD?

**Comment by dteney** [comment]
You're right, but the 0.5 scale factor for the loss is not really relevant. You do want to sum it up over image pixels too, as Gil does.

**Comment by GilLevi** [comment]
Hi,

Thanks for all the comments. As input data I use images that underwent down-sampling and upsampling. The labels are the original images (without downsampling and upsamping). 

@dteney  - thanks for your comments. However, since the first conv layer has stride=2, it makes the output images 2 times smaller than the labels (ground truth high resolution images), so I changed it to stride=1.

Here is my full code:

M=5000;
data=zeros(140,140,1,M);
%gathering data
data=data/255;
imdb.images.data = imresize(imresize(data,0.25),4);
imdb.images.data_mean = mean(data,4);
imdb.images.labels = data ;
imdb.images.set=[ones(1,48*M/50), 2*ones(1,1*M/50) , 3*ones(1,1*M/50)];
imdb.images.data = single(imdb.images.data);
imdb.images.labels = single(imdb.images.labels);


lr=0.001;
initialBias=0;
filters_num1=96;
filters_num2=48;

net.layers{end+1} = struct('type', 'conv', 'filters', 0.01 * randn(9, 9, 1, filters_num1, 'single'), 'biases', zeros(1, filters_num1, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
net.layers{end+1} = struct('type', 'relu') ;

net.layers{end+1} = struct('type', 'conv', 'filters', 0.01 * randn(1, 1, filters_num1, filters_num2, 'single'), 'biases', initialBias*ones(1, filters_num2, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
net.layers{end+1} = struct('type', 'relu') ;

net.layers{end+1} = struct('type', 'conv', 'filters', 0 * randn(5, 5, filters_num2, 1, 'single'), 'biases', initialBias*ones(1, 1, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;

net.layers{end+1} = struct('type', 'RMSEloss') ;


opts.train.batchSize = 10 ;
opts.train.numEpochs = 10 ;
opts.train.continue = false ;
opts.train.useGpu = false ;
opts.train.learningRate = [0.001*ones(1, 12) 0.0001*ones(1,6) 0.00001] ;


[net, info] = cnn_train(net, imdb, @getBatch, ...
    opts.train, ...
    'val', find(imdb.images.set == 2)) ;



% Testing results
test_set=imdb.images.data(:,:,:,(imdb.images.set==3));
test_label=imdb.images.labels(:,:,:,(imdb.images.set==3));
test_set=test_set(:,:,:,1:10);
test_label=test_label(:,:,:,1:10);
stagedPredictions = vl_simplenn(net, test_set);
y_hat = stagedPredictions(end-1).x;



my_err=(test_label-y_hat).^2;
my_err=sum(my_err(:));


init_err=(test_label- test_set).^2;
init_err=sum(init_err(:));

disp(sprintf('my err is %f and init_err is %f',my_err,init_err));



>>my err is 184.554260 and init_err is 110.280609




When testing, I take 10 low resolution images and run the net on them. It seems that the output images of the net actually produce more error than just taking the images before running the net on them. So what am I missing here? why does the net produce such bad results?

Thanks!
Gil

**Comment by bazilas** [comment]
Hi Gil,

You estimate the mean in "imdb.images.data_mean = mean(data,4);". I don't see where you subtract the mean from the training and testing data. Do you have it inside your get_batch function?


Vasilis

**Comment by dteney** [comment]
Two more things on the network initialization:

- It may be easier to train it with a zero-mean output, i.e. yo you compute and substract the mean from the output juste like you do for the input image (in this special case you could probably use the same mean since the input and output are the same kind of images).

- The last (convolution) layer before the loss will most likely converge to small biases, which you should thus initialize to 0. They are usually initialized to some positive value when they are followed by a RELU rectification (so that there is something that actually goes "through" the RELU in the first SGD iterations).

**Comment by GilLevi** [comment]
Hi, 

Thanks for the advice @bazilas,  @dteney and @pengsun . Here is my updated code, the predictions on the test set are still not closer to the ground truth than the test set itself. I wonder what I might be doing wrong? I would be very grateful for some pointers. Thanks!




data=data/255;
imdb.images.data = imresize(imresize(data,0.25),4);
imdb.images.data_mean = mean(data,4);
imdb.images.labels = data ;
imdb.images.set=[ones(1,48*M/50), 2*ones(1,1*M/50) , 3*ones(1,1*M/50)];

imdb.images.data = single(imdb.images.data);
imdb.images.data_mean = single(imdb.images.data_mean);
imdb.images.labels = single(imdb.images.labels);
imdb.images.data = bsxfun(@minus, imdb.images.data, imdb.images.data_mean ) ;
imdb.images.labels = bsxfun(@minus, imdb.images.labels, imdb.images.data_mean ) ;




lr=0.001;
initialBias=0.001;
filters_num1=96;
filters_num2=48;

% Define network architecture
net.layers = {} ;


k1=0.01;

net.layers{end+1} = struct('type', 'conv', 'filters', k1 * randn(9, 9, 1, filters_num1, 'single'), 'biases', zeros(1, filters_num1, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
net.layers{end+1} = struct('type', 'relu') ;

net.layers{end+1} = struct('type', 'conv', 'filters', k1 * randn(1, 1, filters_num1, filters_num2, 'single'), 'biases', initialBias*ones(1, filters_num2, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;
net.layers{end+1} = struct('type', 'relu') ;

net.layers{end+1} = struct('type', 'conv', 'filters', k1 * randn(5, 5, filters_num2, 1, 'single'), 'biases', initialBias*zeros(1, 1, 'single'), 'stride', 1, 'pad', 2, 'filtersLearningRate', lr, 'biasesLearningRate', lr, 'filtersWeightDecay', 1, 'biasesWeightDecay', 0) ;

net.layers{end+1} = struct('type', 'RMSEloss') ;
net_init=net;

opts.train.batchSize = 10 ;
opts.train.numEpochs = 10 ;
opts.train.continue = false ;
opts.train.useGpu = false ;
opts.train.learningRate = [0.001*ones(1, 12) 0.0001*ones(1,6) 0.00001] ;


[net, info] = cnn_train2(net, imdb, @getBatch, ...
    opts.train, ...
    'val', find(imdb.images.set == 2)) ;



test_set=imdb.images.data(:,:,:,(imdb.images.set==3));
test_label=imdb.images.labels(:,:,:,(imdb.images.set==3));
test_set=test_set(:,:,:,1:10);
test_label=test_label(:,:,:,1:10);


stagedPredictions = vl_simplenn(net, test_set);
y_hat = stagedPredictions(end-1).x;

figure; imshow(y_hat(:,:,:,1) + imdb.images.data_mean);title('sr');
figure; imshow(test_set(:,:,:,1) + imdb.images.data_mean);title('data');
figure; imshow(test_label(:,:,:,1)+ imdb.images.data_mean);title('gt');

my_err=(test_label-y_hat).^2;
my_err=sum(my_err(:));


init_err=(test_label - test_set).^2;
init_err=sum(init_err(:));

disp(sprintf('my err is %f and init_err is %f',my_err,init_err));




And the output is:
my err is 203.102295 and init_err is 110.280594

**Comment by GilLevi** [comment]
OK, I found my bug. Just a matter of letting it run with a few more epochs. 

I'll make some more tests with various parameters to be completely sure that everything is working and then I'll close the thread.

Thanks again to  @bazilas, @dteney and @pengsun !!

**Comment by atique81** [comment]
Hi @vedaldi, @dteney, @bazilas, @GilLevi 

I am also trying to implement a customized loss function which is the squared hinge-loss function, also known as the L2-SVM loss as follows.

function [loss, dzdx] = squaredHingeLoss(X, labels, dzdy)
a = squeeze(X);
a = max(0, 1-a.*labels)^2;
loss = sum(a); % summing over all the images in a batch
if nargin<=2 retrun; end

dzdx = 2*a*(-labels);
dzdx = reshape(dzdx, size(X));

My question is, do I need to divide the dzdx by the batch size like this ?? --> dzdx = dzdx/size(X, 4)
